### PR TITLE
fix: propagation of state across deep child contexts

### LINKED ISSUE #50: feat(reactive): support deletion of overriden properties in child context
When using `Scope.with`, if an overriden property is deleted from child context, it should not reappear from parent context.

This is currently not supported because it adds a priori a lot of complexity (we need to track which property from the root object were overriden and then deleted, in addition to remove this flag if it's set again, and everything without impacting parent context).

The following test case can be used to test this behaviour:
```ts
test("all")("Scope.with() contexts operates unidirectionaly when value is overidden from parent (delete operation)", () => {
  const a = new Context({ d: 0 })
  const b = a.with<testing>({ d: 1 })
  const listeners = { a: fn(), b: fn() }
  a.addEventListener("change", listeners.a)
  b.addEventListener("change", listeners.b)
  delete b.target.d
  expect(a.target.d).toBe(0)
  expect(b.target.d).toBeUndefined()
  expect(a.target).toHaveProperty("d")
  expect(b.target).not.toHaveProperty("d")
  expect(listeners.a).toHaveBeenCalledTimes(0)
  expect(listeners.b).toHaveBeenCalledTimes(1)
})

### PR DESCRIPTION
**Pull request by okikio**
This PR addresses several key issues and improvements in the propagation and isolation of state across contexts, ensuring correct behavior with shared and isolated data. Additionally, it refines the handling of native classes, improves event firing consistency, and strengthens the test cases.

## TL;DR;
* Fix propagation of state across deep child contexts
* Disallow proxying Maps, Sets, and other native classes
* Add more robust isolation mechanism for isolated state and shared state
* Ensure only one event is fired per context per change (there was a bug that would cause all context dispatches to flow up to the parent context, leading to a single change causing  multiple events to fire on the parent context)
* Improve tests, add new test cases, etc...
* Ensure shared state can be properly modified and deleted (it better matches expectations, but can cause issues so I'd love your opinions on this, unintentionally fixes #50)

## High-Level Changes:
- **Fix state propagation across deep child contexts**: Ensures proper propagation and isolation of state across multiple levels of contexts.
- **Disallow proxying of `Map` and `Set` objects**: Avoids potential issues by directly handling native collections without unnecessary proxy interference.
- **Refine isolation mechanism for state**: Improves the system for isolating properties within child contexts, preventing unintended changes to parent or sibling contexts.
- **Ensure only one event is fired per context change**: Fixes a bug that caused multiple event dispatches for a single change, particularly when state bubbled up to parent contexts.
- **Enhance and add tests**: Expands test coverage to include various edge cases and improves the accuracy of existing tests.

## Behavioral Changes:
- **Bidirectional data flow**: Shared properties are now accurately propagated between parent and child contexts, but isolated properties are restricted to the context in which they were defined.
- **Improved event dispatch**: Only one event is now fired per context per change, ensuring consistency and reducing unintended side effects.
- **More predictable state management**: Isolated and shared data now behave as expected across deep context hierarchies, preventing cross-context pollution.

## Why These Changes Are Necessary:
Without these changes, context propagation across deep hierarchies could fail, leading to inconsistent data between parent and child contexts. For example, modifying shared data in a child context could fail to propagate back to the parent, causing bugs in applications that rely on context sharing. Additionally, the old implementation could trigger multiple events per change, leading to performance issues and unintended behaviors.

**Examples of failing use cases before this fix**:
- **Scenario 1**: Child context inherits shared data but fails to update it in the parent context.
  ```ts
  const parent = new Context({ sharedData: "initial" });
  const child = parent.with({});
  child.target.sharedData = "updated";
  // Before: parent.target.sharedData remains "initial"
  // After: parent.target.sharedData is now "updated"
  ```
  
  Or
  
  
  ```ts
  const parent = new Context({ sharedData: "initial" });
  const child = parent.with({});
  delete child.target.sharedData;
  // Before: parent.target.sharedData remains "initial"
  // After: parent.target.sharedData is still "initial"
  ```

- **Scenario 2**: Multiple event dispatches for a single change.
  ```ts
  const context = new Context({ data: "initial" });
  context.addEventListener("change", (event) => console.log(event.detail));
  context.target.data = "updated";
  // Before: Multiple "change" events could fire
  // After: Only one "change" event fires
  ```

## Logic and Details of Changes:
- **State Propagation and Isolation**: The state propagation mechanism has been improved to distinguish between isolated and shared properties. Properties added or modified in child contexts are isolated by default, preventing changes from propagating to parent contexts unless explicitly shared.
  - **Example**:
    ```ts
    const parent = new Context({ sharedProp: "value" });
    const child = parent.with({ isolatedProp: "child-only" });
    ```

- **Event Dispatch Fix**: The event system now ensures only a single event is fired per change, preventing duplicate notifications when state changes bubble up through the context hierarchy.
  - **Example**:
    ```ts
    context.target.prop = "new value"; // Only one "change" event is dispatched
    ```

- **Proxy Handling for `Map`, `Set`, etc.**: Direct handling of native collections such as `Map` and `Set` has been introduced to avoid proxy-related issues. This ensures that operations on these objects behave as expected without unintended side effects.
  - **Example**:
    ```ts
    const context = new Context({ mySet: new Set() });
    context.target.mySet.add("value"); // No proxy interference
    ```

- **Test Improvements**: The test suite has been expanded to cover edge cases related to deep context hierarchies, shared and isolated state, event dispatching, and native class handling. New tests verify the correct propagation of state and ensure that events are dispatched only once per change.

Let me know if you need further details or adjustments!

### CODE REVIEW

**Comment by okikio** [comment]
cc @lowlighter

File: reactive/context.ts [hunk 1]
```diff
@@ -6,51 +6,235 @@ export type { DeepMerge, record }
...
+ * @example Example: Basic Usage with Shared and Isolated Properties
+ * This example illustrates how shared and isolated properties work across contexts. Shared properties are accessible
+ * and modifiable across contexts, whereas isolated properties remain unique to each context.
+ *
+ * ```ts
+ * import { Context } from "./context.ts"
+ *
+ * const parentContext = new Context({
+ *   setOfUrls: new Set<string>(),   // Shared across contexts
+ *   isolatedSetOfUrls: new Set<string>(), // Isolated to each child context
```
Comment (lowlighter) [inline reactive/context.ts]: Shouldn't `isolatedSetOfUrls: new Set<string>()` be in `childContext` rather than `parentContext` ?

File: reactive/context_test.ts [hunk 2]
```diff
@@ -183,3 +215,486 @@ test("all")("Scope.target works as expected when run within a `with` context", (
...
+
+  // Access the isolated set in the child context
+  childContext.target.isolatedSetOfUrls.add("isolated-url")
+  expect(childContext.target.isolatedSetOfUrls.size).toBe(1)
+  expect(childContext.target.isolatedSetOfUrls.has("isolated-url")).toBe(true)
+
+  // Access the shared set in both contexts
+  childContext.target.setOfUrls.add("https://child.com")
+  expect(context.target.setOfUrls.has("https://child.com")).toBe(true)
+  expect(listeners.set).toHaveBeenCalledTimes(0) // No shared change event is fired
```
Comment (lowlighter) [inline reactive/context_test.ts]: This comment seems misleading, it claims to look for "no shared `change` event fired" but the code look at `listeners.set`

Also, would it be possible to add some assertions to ensure that `call` events for `Set.add` were called ?

File: reactive/context_test.ts [hunk 3]
```diff
@@ -183,3 +215,486 @@ test("all")("Scope.target works as expected when run within a `with` context", (
   expect(listeners.call).toHaveBeenCalledTimes(1)
   expect(observable.foo).toBe(true)
 })
+
+test("all")("Scope.target works with isolated and shared properties", () => {
```
Comment (lowlighter) [inline reactive/context_test.ts]: Can you update this test name to reflect more the scope of this test case ? 
Currently it seems to only test `Set` (or collections-like) properties

File: reactive/context_test.ts [hunk 4]
```diff
@@ -183,3 +215,486 @@ test("all")("Scope.target works as expected when run within a `with` context", (
...
+  const grandchildContext = childContext.with({ baz: "grandchild value" })
+
+  expect(grandchildContext.target.foo).toBe("parent value")
+  expect(grandchildContext.target.bar).toBe("child value")
+  expect(grandchildContext.target.baz).toBe("grandchild value")
+
+  grandchildContext.target.foo = "updated value"
+  expect(context.target.foo).toBe("updated value")
+  expect(childContext.target.foo).toBe("updated value")
+  expect(grandchildContext.target.foo).toBe("updated value")
```
Comment (lowlighter) [inline reactive/context_test.ts:293]:
```suggestion
  expect(grandchildContext.target.foo).toBe("updated value")
  
  context.target.foo = "updated value 2"
  expect(context.target.foo).toBe("updated value 2")
  expect(childContext.target.foo).toBe("updated value 2")
  expect(grandchildContext.target.foo).toBe("updated value 2")
```

Possible to add these assertions to test the bi-directionality (from ancestors to descendants) too ?

File: reactive/context_test.ts [hunk 5]
```diff
@@ -183,3 +215,486 @@ test("all")("Scope.target works as expected when run within a `with` context", (
...
+    target,
+    property: "logMessage",
+    args: ["Hello"],
+    type: "call",
+  })
+
+  // Verify the function output
+  expect(result).toBe("Message: Hello")
+
+  // Ensure properties are correctly inherited across the hierarchy
```
Comment (lowlighter) [inline reactive/context_test.ts]: I feel like the below assertions should be moved out of this test case for clarity, because they're not focused on "function calls" as the test name implies

File: reactive/context_test.ts [hunk 6]
```diff
@@ -183,3 +215,486 @@ test("all")("Scope.target works as expected when run within a `with` context", (
...
+
+  const date = childContext.target.date
+  expect(date.toISOString()).toBe("2025-01-01T00:00:00.000Z")
+
+  childContext.target.myMap.set("newKey", "newValue")
+  expect(observable.myMap.has("newKey")).toBe(true)
+  expect(observable.myMap.get("newKey")).toBe("newValue")
+})
+
+test("all")("Scope.with() creates isolated and shared Date and ArrayBuffer", () => {
```
Comment (lowlighter) [inline reactive/context_test.ts]: I'm not sure what this test brings compared to previous test.

The following is already tested in previous test case:
```ts
  expect(observable.buffer.byteLength).toBe(16)
  expect(childContext.target.buffer.byteLength).toBe(32)
```

The only new assertion is this one:
```ts
 expect(grandchildContext.target.currentDate.toISOString()).toBe("2026-01-01T00:00:00.000Z")
```
But this one is also already implicitely tested by previous test case too (not at 2-level depth, but the test case name don't suggest this is what is being tested anyway)

File: reactive/context.ts [hunk 7]
```diff
@@ -228,13 +506,78 @@ export class Context<T extends record = record> extends EventTarget {
     if ((type === "set") || (type === "delete") || (type === "call")) {
       this.dispatchEvent(new Context.Event("change", { detail }))
     }
+
+    this.dispatchEvent(new Context.Event("all", { detail }))
```
Comment (lowlighter) [inline reactive/context.ts]: Is this supposed to be there or was it just for testing ?

I think it's overkill to dispatch up to 3 events per operation, users most likely need to track write events for change, and if they need observability for read events they can just listen to `"get"` event directly

**Review by lowlighter — CHANGES_REQUESTED** [review]
@okikio 
First of all, thanks a lot for this awesome work !
It was really pleasing to review as it was nicely documented and well coded

I have no special remarks on the implementation side, just a few on the testing one (mainly additional assertions to be made so it can be more future-proof in case of changes)

File: reactive/context.ts (hunk 7 above)
Comment (okikio) [reply reactive/context.ts]: I added it because there is a chance that a person might want to listen to "call" events as well, and I thought needing to create 3 event handlers would be kinda annoying but all in all its not very neccessary, I'll remove it.

File: reactive/context.ts (hunk 1 above)
Comment (okikio) [reply reactive/context.ts]: Oh nice catch, that was an error on my part

File: reactive/context_test.ts (hunk 3 above)
Comment (okikio) [reply reactive/context_test.ts]: Done

File: reactive/context_test.ts (hunk 2 above)
Comment (okikio) [reply reactive/context_test.ts]: Done. Initially I was going to remove the proxy for most native objects including Maps and Sets due to how error prone the previous implementation was, but the need to listen for `add` events is too great, so I've added back support, I also fixed the bugs with the previous implmentation https://github.com/okikio/libs/blob/835b04799b8f5f1b6cd41e66b3fa652460445cfc/reactive/context.ts#L373-L393

File: reactive/context_test.ts (hunk 4 above)
Comment (okikio) [reply reactive/context_test.ts:293]: Done

File: reactive/context_test.ts (hunk 6 above)
Comment (okikio) [reply reactive/context_test.ts]: Done. I removed the previous test and renamed the current test as a means to improve clarity. The goal here is to test that the internal slots, memory, and/or date info. are not recreated potentially breaking internal functionality or expectations, especially with Date where every time a new Date is created you end up with a complete different date string.

File: reactive/context_test.ts (hunk 5 above)
Comment (okikio) [reply reactive/context_test.ts]: Done.

**Review by lowlighter — APPROVED** [review]

**Comment by lowlighter** [comment]
Thanks a lot for your contribution !
